File: src/MOC.m
% 假设阀门瞬间关闭
clear
tic
L=300;                          % 管线长度
Hr=70;                          % 泵压力
N=10;                           % 分段数
NS=N+1;                         % 节点数
e=0.001651;                     % 壁厚m,0.065''
D=0.00635-2*e;	                % 管道内径
K=2.1e+9;                       % 流体体积弹性系数
Rho=1000;	                    % 液体密度kg/m^3
E=2.1e11;                       % 弹性模数
G=9.806;                        % 重力加速度
f=0.018;                        % 摩擦系数
A=pi*(D^2)/4;                     % 管道横截面积

% 网格划分
dx=L/N;                         % 单位长度
t_max=20;                       % 总计算时间
a=sqrt(K/Rho/(1+K*D/(E*e)));     % 波速
dt=dx/a;                        % 单位计算时间

k=1;
t=0;time(k)=t;

% 常数B和R
B=a/(G*A);
R=f*dx/(2*G*D*(A^2));

% 第一列流量和压力
u0=0.1; % initial volecity

Q(k,1:NS)=0; % A*u0;
H(k,2:NS)=0;
H(1,1)=Hr;

while t<t_max
   for j=2:N
        CP=H(k,j-1)+B*Q(k,j-1)-R*Q(k,j-1)*abs(Q(k,j-1));    % +dt*Q(k,j-1)/A;    % CP==C-==CR
        CM=H(k,j+1)-B*Q(k,j+1)+R*Q(k,j+1)*abs(Q(k,j+1));    % -dt*Q(k,j+1)/A;    % CM==C+==CL
        H(k+1,j)=(CP+CM)/2;                                 % 计算压头
        Q(k+1,j)=(H(k+1,j)-CM)/(B);                         % 计算流量
         
   end
    
    % 边界值
    CP=H(k,N)+B*Q(k,N)-R*Q(k,N)*abs(Q(k,N));    % +dt*Q(k,N)/A;    % 单独计算CP
    CM=H(k,2)-B*Q(k,2)+R*Q(k,2)*abs(Q(k,2));    % -dt*Q(k,2)/A;    % 单独计算CM

    H(k+1,1)=Hr;                                % 压头第一列
    Q(k+1,1)=(H(k+1,1)-CM)/B;                   % 流量第一列数值，边界条件公式
    
    Q(k+1,NS)=0;                                % 流量最后一列
    H(k+1,NS)=CP;                               % 压头最后一列数值，边界条件公式

    t=t+dt;
    k=k+1;
    time(k)=t;
    
end
toc

plot(time,H(:,N+1))
title('MOC-阀门处圧力曲线');
xlabel('单位：s');
ylabel('单位：m');

File: src/MOC.py
import time as timer
import numpy as np
import matplotlib.pyplot as plt


def simulate_instant_valve_closure():
    """
    Simulate water hammer in a pipeline using the Method of Characteristics (MOC)
    under the assumption of instantaneous valve closure at the downstream end.

    Main fixes compared with the original MATLAB script:
    1. The initial steady-state head profile is made physically consistent.
    2. Boundary conditions are computed from the previous time level, which is the
       correct MOC update for this formulation.
    3. The reference head line is plotted with the correct 1D shape.
    """

    # Pipe and fluid parameters
    L = 300.0                    # Pipe length (m)
    Hr = 70.0                    # Reservoir / pump head (m)
    N = 10                       # Number of pipe reaches
    NS = N + 1                   # Number of nodes
    e = 0.001651                 # Wall thickness (m), 0.065 in
    D = 0.00635 - 2.0 * e        # Inner diameter (m)
    K = 2.1e9                    # Bulk modulus (Pa)
    rho = 1000.0                 # Fluid density (kg/m^3)
    E = 2.1e11                   # Young's modulus (Pa)
    g = 9.806                    # Gravitational acceleration (m/s^2)
    f = 0.018                    # Darcy-Weisbach friction factor (-)
    area = np.pi * D**2 / 4.0    # Cross-sectional area (m^2)

    # Grid and time step
    dx = L / N                   # Spatial step (m)
    t_max = 20.0                 # Total simulation time (s)
    a = np.sqrt(K / rho / (1.0 + K * D / (E * e)))  # Wave speed (m/s)
    dt = dx / a                  # Time step from the Courant condition (s)

    # MOC constants
    B = a / (g * area)
    R = f * dx / (2.0 * g * D * area**2)

    # Initial steady flow
    u0 = 0.1                     # Initial velocity (m/s)

    # Preallocate arrays
    n_steps = int(np.ceil(t_max / dt)) + 1
    time = np.zeros(n_steps)
    Q = np.zeros((n_steps, NS))
    H = np.zeros((n_steps, NS))

    # Initial discharge at every node
    Q[0, :] = area * u0

    # Initial steady-state head profile along the pipe
    # x = 0 at the upstream reservoir, x = L at the downstream valve
    x = np.arange(NS) * dx
    H[0, :] = Hr - f * x * u0**2 / (2.0 * g * D)
    H[0, 0] = Hr

    last_index = 0
    for k in range(n_steps - 1):
        # Interior nodes: Python indices 1..N-1 correspond to MATLAB 2..N
        for j in range(1, N):
            CP = H[k, j - 1] + B * Q[k, j - 1] - R * Q[k, j - 1] * abs(Q[k, j - 1])
            CM = H[k, j + 1] - B * Q[k, j + 1] + R * Q[k, j + 1] * abs(Q[k, j + 1])

            H[k + 1, j] = 0.5 * (CP + CM)
            Q[k + 1, j] = (H[k + 1, j] - CM) / B

        # Upstream boundary: constant head reservoir
        # Use the previous time level for the characteristic coming from node 2.
        CM = H[k, 1] - B * Q[k, 1] + R * Q[k, 1] * abs(Q[k, 1])
        H[k + 1, 0] = Hr
        Q[k + 1, 0] = (H[k + 1, 0] - CM) / B

        # Downstream boundary: instantaneous valve closure, so Q = 0
        # Use the previous time level for the characteristic coming from node N.
        CP = H[k, N - 1] + B * Q[k, N - 1] - R * Q[k, N - 1] * abs(Q[k, N - 1])
        Q[k + 1, N] = 0.0
        H[k + 1, N] = CP

        time[k + 1] = time[k] + dt
        last_index = k + 1

        if time[k + 1] >= t_max:
            break

    # Trim arrays to the actual simulated length
    time = time[: last_index + 1]
    Q = Q[: last_index + 1, :]
    H = H[: last_index + 1, :]

    return time, H, Q, dt, a


def main():
    start = timer.perf_counter()
    time, H, Q, dt, a = simulate_instant_valve_closure()
    elapsed = timer.perf_counter() - start

    print(f"Computed wave speed a = {a:.6f} m/s")
    print(f"Time step dt = {dt:.6e} s")
    print(f"Number of time steps = {len(time)}")
    print(f"Elapsed time = {elapsed:.6f} s")

    plt.figure(figsize=(8, 5))
    plt.plot(time, H[:, -1], label="Head at the closed valve")
    plt.plot(time, np.full_like(time, 70.0), "--", label="Reservoir head")
    plt.title("MOC Pressure Head Curve at the Valve")
    plt.xlabel("Time (s)")
    plt.ylabel("Head (m)")
    plt.grid(True)
    plt.legend()
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    main()